# Exploring Ebay Car Sales Data

In this project we will work with this [dataset](https://www.kaggle.com/orgesleka/used-cars-database/data) of used cars from *eBay Kleinanzeigen*, a classifieds section of the German eBay website.

The dataset, originally uploaded in *Kaggle*, was modified by sampling 50,000 data points from the full dataset and by making it more "dirty" so it would resemble better what would be expected from a scraped dataset.

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

With this dataset, we will clean the data and analyse the included used car listings.

## The Dataset

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')



In [2]:
autos.info()
autos.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

From this primary inspection, we can observe that:
- our dataset has some missing values for some columns (NaN fields)
- Values as `price` and `odometerer` are registered as an object(string) and we will have to change its type to integer to analyse it
- The column names use camelcase instead of Python's preferred snake_case, which means we can't just replace spaces with underscores


## Changing the fields names 

We will convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [3]:
columns_mapper = {
    'dateCrawled' : 'date_crawled',
    'offerType' : 'offer_type',
    'vehicleType' : 'vehicle_type',
    'yearOfRegistration' : 'registration_year',
    'monthOfRegistration' : 'registration_month',
    'fuelType' : 'fuel_type',
    'notRepairedDamage' : 'unrepaired_damage',
    'dateCreated': 'ad_created',
    'nrOfPictures' : 'nr_of_pictures',
    'postalCode' : 'postal_code',
    'lastSeen' : 'last_seen'
}

autos.rename(columns = columns_mapper, inplace=True)
autos.head()


    

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning

In [4]:
autos.describe(include= 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From the description above, we can infer that there are some data to be cleaned, changed or more investigated. For instance:

- `price` stored as strings or nil values  (hence the top being `$0`)
- `odometer` stored as string
- columns that will not be of our interest since the values remains the same for all the data such as `seller`, `offer_type` and `nr_of_picures`
- `registration_year` column having max set as 9999.00000
- `postal_code` as float is not interesting to analyse as it is actually not countable
- turn all columns related to data and time in `datetime` object

First, we will handle the `odometer` and `price` columns by removing any non numeric characters and turning the values in integers/floats

In [5]:
autos["odometer"] = autos['odometer'].str.replace('km','')
autos["odometer"] = autos['odometer'].str.replace(',','')
autos["price"] = autos['price'].str.replace('$','')
autos["price"] = autos['price'].str.replace(',','')


In [6]:
autos["price"] = autos["price"].astype(float)
autos["odometer"] = autos["odometer"].astype(int)
autos.rename(columns = {"odometer":"odometer_km"}, inplace=True)


In [7]:
autos.describe(include= 'all')
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  5000.0  control          bus               2004    manuell   
1    Angebot  8500.0  control    limousine               1997  automatik   
2    Angebot  8990.0     test    limousine               2009    manuell   

   powerPS   model  odometer_km  registration_month fuel_type       brand  \
0      158  andere       150000                   3       lpg     peugeot   
1      286     7er       150000                   6    benzin         bmw   
2      102    golf        70000                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

## Looking for and removing outliers

Let's analyse `odometer_km` and `price` columns:

In [8]:
print ("odometer_km and prices have this many unique values \n")

print("odometer_km: ",autos["odometer_km"].unique().shape)
print("price: ",autos["price"].unique().shape, "\n")

print ("About basic statistics: \n")
print (autos["odometer_km"].describe(), "\n")
print (autos["price"].describe(), "\n")
print ("For frequency of values: \n")
print (autos["odometer_km"].value_counts().sort_index(ascending=True),"\n")
print (autos["price"].value_counts().sort_index(ascending=False).head(10), "\n")
print (autos["price"].value_counts().sort_index(ascending=True).head(10), "\n")


odometer_km and prices have this many unique values 

odometer_km:  (13,)
price:  (2357,) 

About basic statistics: 

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

For frequency of values: 

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64 

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
Name: price, dtype: int64 


In [9]:
autos = autos[autos["price"].between(1,1000000)]
autos["price"].value_counts().sort_index()


1.0         156
2.0           3
3.0           1
5.0           2
8.0           1
           ... 
299000.0      1
345000.0      1
350000.0      1
999990.0      1
999999.0      2
Name: price, Length: 2348, dtype: int64

For `price` column, cars with no price and price above $1,000,000 were removed.
As for `odometer_km`, no outliers were found

## Handling Date Types

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values. Let's first understand how the values in the three string columns are formatted:


In [10]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We will then extract the date from those strings by selecting the first 10 characters of each date data (which represents the date itself), and explore the three columns.

#### For `date_crawled` :

In [11]:
date_crawled = autos["date_crawled"].str[:10]
date_crawled.value_counts(normalize=True, dropna=False)

2016-04-03    0.038606
2016-03-20    0.037885
2016-03-21    0.037391
2016-03-12    0.036917
2016-03-14    0.036547
2016-04-04    0.036485
2016-03-07    0.036011
2016-04-02    0.035476
2016-03-28    0.034858
2016-03-19    0.034776
2016-03-15    0.034282
2016-03-29    0.034117
2016-03-30    0.033685
2016-04-01    0.033685
2016-03-08    0.033294
2016-03-09    0.033088
2016-03-22    0.032985
2016-03-11    0.032573
2016-03-23    0.032223
2016-03-26    0.032202
2016-03-10    0.032182
2016-03-31    0.031832
2016-03-17    0.031646
2016-03-25    0.031605
2016-03-27    0.031090
2016-03-16    0.029608
2016-03-24    0.029340
2016-03-05    0.025325
2016-03-13    0.015669
2016-03-06    0.014042
2016-04-05    0.013095
2016-03-18    0.012910
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

We can observe that the ads were first crawled in March and April of 2016, specially from mid March.

#### For `ad_created` :

In [12]:
ad_created = autos["ad_created"].str[:10]
ad_created.value_counts(normalize=True, dropna=False)



2016-04-03    0.038853
2016-03-20    0.037947
2016-03-21    0.037597
2016-04-04    0.036856
2016-03-12    0.036753
                ...   
2016-02-01    0.000021
2015-09-09    0.000021
2015-08-10    0.000021
2016-01-29    0.000021
2016-02-08    0.000021
Name: ad_created, Length: 76, dtype: float64

In [13]:
ad_created.value_counts(normalize=True, dropna=False).head(30)

2016-04-03    0.038853
2016-03-20    0.037947
2016-03-21    0.037597
2016-04-04    0.036856
2016-03-12    0.036753
2016-03-14    0.035188
2016-04-02    0.035147
2016-03-28    0.034982
2016-03-07    0.034735
2016-03-29    0.034055
2016-03-15    0.034014
2016-04-01    0.033685
2016-03-19    0.033685
2016-03-30    0.033499
2016-03-08    0.033314
2016-03-09    0.033149
2016-03-11    0.032902
2016-03-22    0.032799
2016-03-26    0.032264
2016-03-23    0.032058
2016-03-10    0.031893
2016-03-31    0.031873
2016-03-25    0.031749
2016-03-17    0.031296
2016-03-27    0.030987
2016-03-16    0.030123
2016-03-24    0.029279
2016-03-05    0.022896
2016-03-13    0.017007
2016-03-06    0.015319
Name: ad_created, dtype: float64

The ads were first crawled by the same time they were created the most (March and April 2016). 


#### For `last_seen` :

In [14]:
last_seen = autos["last_seen"].str[:10]
last_seen.value_counts(normalize=True, dropna=False)

2016-04-06    0.221813
2016-04-07    0.131939
2016-04-05    0.124753
2016-03-17    0.028084
2016-04-03    0.025202
2016-04-02    0.024914
2016-03-30    0.024769
2016-04-04    0.024481
2016-03-12    0.023781
2016-03-31    0.023781
2016-04-01    0.022793
2016-03-29    0.022360
2016-03-22    0.021372
2016-03-28    0.020878
2016-03-20    0.020651
2016-03-21    0.020631
2016-03-24    0.019766
2016-03-25    0.019210
2016-03-23    0.018531
2016-03-26    0.016801
2016-03-16    0.016451
2016-03-15    0.015875
2016-03-19    0.015833
2016-03-27    0.015648
2016-03-14    0.012601
2016-03-11    0.012374
2016-03-10    0.010665
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007412
2016-03-18    0.007351
2016-03-07    0.005394
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

In [15]:
last_seen.value_counts(normalize=True, dropna=False).head(30)

2016-04-06    0.221813
2016-04-07    0.131939
2016-04-05    0.124753
2016-03-17    0.028084
2016-04-03    0.025202
2016-04-02    0.024914
2016-03-30    0.024769
2016-04-04    0.024481
2016-03-12    0.023781
2016-03-31    0.023781
2016-04-01    0.022793
2016-03-29    0.022360
2016-03-22    0.021372
2016-03-28    0.020878
2016-03-20    0.020651
2016-03-21    0.020631
2016-03-24    0.019766
2016-03-25    0.019210
2016-03-23    0.018531
2016-03-26    0.016801
2016-03-16    0.016451
2016-03-15    0.015875
2016-03-19    0.015833
2016-03-27    0.015648
2016-03-14    0.012601
2016-03-11    0.012374
2016-03-10    0.010665
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007412
Name: last_seen, dtype: float64

The same applies for `last_seen` 

#### For `registration_year`

In [16]:
autos["registration_year"].describe()

count    48568.000000
mean      2004.754612
std         88.641262
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

For this column, we can observe that are values that are faulty, for instance `1000`, `9999` and those 25% that are dates from `1000` up to `1999`. Furthermore, because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. We will then consider reliable registration years between 1960 and 2016. (We are not interested in really antiques, are we?)



In [17]:
autos = autos[autos["registration_year"].between(1960,2016)]
autos["registration_year"].value_counts()

2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
2006    2670
2001    2636
2002    2486
1998    2363
2007    2277
2008    2215
2009    2086
1997    1951
2011    1623
2010    1589
1996    1373
2012    1310
1995    1227
2016    1220
2013     803
2014     663
1994     629
1993     425
2015     392
1992     370
1990     347
1991     339
1989     174
1988     135
1985      96
1980      85
1986      72
1987      72
1984      51
1983      51
1978      44
1982      41
1970      38
1979      34
1972      33
1981      29
1968      26
1967      26
1971      26
1974      24
1960      23
1973      23
1966      22
1977      22
1976      21
1969      19
1975      18
1965      17
1964      12
1963       8
1961       6
1962       4
Name: registration_year, dtype: int64

Thus, most cars are registered from the 90's decade onwards. Let's analyse in which decade (from 90's) falls the higher number of registered cars.

In [18]:
autos_90 = autos[autos["registration_year"].between(1990,1999)]
autos_00 = autos[autos["registration_year"].between(2000,2009)]
autos_10 = autos[autos["registration_year"].between(2010,2019)]

In [25]:
print ("From the 90's: ", autos_90.shape[0])
print ("From the 00's: ", autos_00.shape[0])
print ("From the 10's: ", autos_10.shape[0])
print ("Therefore, cars registered in the first decade of 2000 represent ", round((autos_00.shape[0]/autos.shape[0]*100),2), " %")


From the 90's:  11921
From the 00's:  25864
From the 10's:  7600
Therefore, cars registered in the first decade of 2000 represent  55.46  %


## By Brand

We will now analyse the dataset by brand and choose the most representative ones to explore the **mean price for each**. We will be selecting those brands that represent up to 80% of the dataset (**Hello Pareto Principle!**), ranging them from higher to lower.

In [20]:
autos.shape[0]
print ("80% of the dataset is equivalent to ", round(0.8*autos.shape[0],2), " entries")
counts_brand = autos["brand"].value_counts()
counts_brand.cumsum()


80% of the dataset is equivalent to  37309.6  entries


volkswagen         9862
bmw               14998
opel              20016
mercedes_benz     24511
audi              28552
ford              31810
renault           34010
peugeot           35403
fiat              36600
seat              37452
skoda             38217
nissan            38930
mazda             39639
smart             40300
citroen           40953
toyota            41546
hyundai           42014
sonstige_autos    42455
volvo             42882
mini              43291
mitsubishi        43675
honda             44041
kia               44371
alfa_romeo        44680
porsche           44965
suzuki            45242
chevrolet         45507
chrysler          45671
dacia             45794
daihatsu          45911
jeep              46017
subaru            46117
land_rover        46214
saab              46291
jaguar            46364
daewoo            46434
trabant           46498
rover             46560
lancia            46610
lada              46637
Name: brand, dtype: int64

Hence we will consider the following brands:
- volkswagen 
- bmw
- opel
- mercedes_benz
- audi
- ford
- renault
- peugeot
- fiat
- seat

In [21]:
volkswagen_mean_price = round(autos[autos["brand"] == "volkswagen"]["price"].mean(),2)
bmw_mean_price = round(autos[autos["brand"] == "bmw"]["price"].mean(),2)
opel_mean_price = round(autos[autos["brand"] == "opel"]["price"].mean(),2)
mercedes_benz_mean_price = round(autos[autos["brand"] == "mercedes_benz"]["price"].mean(),2)
audi_mean_price = round(autos[autos["brand"] == "audi"]["price"].mean(),2)
ford_mean_price = round(autos[autos["brand"] == "ford"]["price"].mean(),2)
renault_mean_price = round(autos[autos["brand"] == "renault"]["price"].mean(),2)
peugeot_mean_price = round(autos[autos["brand"] == "peugeot"]["price"].mean(),2)
fiat_mean_price = round(autos[autos["brand"] == "fiat"]["price"].mean(),2)
seat_mean_price = round(autos[autos["brand"] == "seat"]["price"].mean(),2)

brand_mean_price = dict (volkswagen=volkswagen_mean_price, bmw=bmw_mean_price, opel=opel_mean_price, mercedes_benz=mercedes_benz_mean_price,
                   audi=audi_mean_price, ford=ford_mean_price, renault=renault_mean_price, peugeot=peugeot_mean_price,
                   fiat=fiat_mean_price, seat=seat_mean_price)

brand_mean_price_s = pd.Series(brand_mean_price)
brand_mean_price_s.sort_values(ascending=False)

audi             9336.69
mercedes_benz    8565.09
bmw              8332.20
volkswagen       5600.57
seat             4402.39
ford             4018.82
peugeot          3094.02
opel             2974.69
fiat             2813.75
renault          2475.72
dtype: float64

Therefore, the most expensive cars are also the ones that have more ads. Besides, we can observe that there is a significant price gap between the top 3 brands (Audi, Mercedes Benz and BMW).

Now let's analyse if the average mileage is related to the mean price (or those are just because of the brand name). 

For this one, we will create a dataframe using Pandas Dataframe constructor with two columns: one for the `mean price` and another one for `mean_mileage`.

In [22]:
volkswagen_mean_mileage = round(autos[autos["brand"] == "volkswagen"]["odometer_km"].mean(),2)
bmw_mean_mileage = round(autos[autos["brand"] == "bmw"]["odometer_km"].mean(),2)
opel_mean_mileage = round(autos[autos["brand"] == "opel"]["odometer_km"].mean(),2)
mercedes_benz_mean_mileage = round(autos[autos["brand"] == "mercedes_benz"]["odometer_km"].mean(),2)
audi_mean_mileage = round(autos[autos["brand"] == "audi"]["odometer_km"].mean(),2)
ford_mean_mileage = round(autos[autos["brand"] == "ford"]["odometer_km"].mean(),2)
renault_mean_mileage = round(autos[autos["brand"] == "renault"]["odometer_km"].mean(),2)
peugeot_mean_mileage = round(autos[autos["brand"] == "peugeot"]["odometer_km"].mean(),2)
fiat_mean_mileage = round(autos[autos["brand"] == "fiat"]["odometer_km"].mean(),2)
seat_mean_mileage = round(autos[autos["brand"] == "seat"]["odometer_km"].mean(),2)

brand_mean_mileage = dict (volkswagen=volkswagen_mean_mileage, bmw=bmw_mean_mileage, opel=opel_mean_mileage, mercedes_benz=mercedes_benz_mean_mileage,
                   audi=audi_mean_mileage, ford=ford_mean_mileage, renault=renault_mean_mileage, peugeot=peugeot_mean_mileage,
                   fiat=fiat_mean_mileage, seat=seat_mean_mileage)

brand_ratio = pd.Series(brand_mean_price)/pd.Series(brand_mean_mileage)
brand_df = pd.DataFrame(data=[brand_mean_price, brand_mean_mileage, brand_ratio], index=['mean_price', 'mean_mileage', 'ratio']).T
brand_df.sort_values('ratio', ascending=False).head(10)

mean_price  mean_mileage     ratio
audi              9336.69     129157.39  0.072289
mercedes_benz     8565.09     130919.91  0.065422
bmw               8332.20     132597.35  0.062838
volkswagen        5600.57     128712.23  0.043512
seat              4402.39     121267.61  0.036303
ford              4018.82     124399.94  0.032306
peugeot           3094.02     127153.63  0.024333
fiat              2813.75     117121.97  0.024024
opel              2974.69     129342.37  0.022999
renault           2475.72     128127.27  0.019322

Therefore, the lower the mileage, more expensive the car is (as expected).Still, we see no actual connection between the average mileage and the mean price, as the most expensive brands still stand in the top list even though there are other brands with lower mileage.

## Wrapping up (and I am not talking about burritos)

We have a dataset that is most represented by:
- Volkswagen, BMW, Mercedes Benz and Audi
- Vehicles from the 00's (millenials everywhere, huh?
- Ads were accessed specially in March and April 2016

Also, I relised (I mean we, right?) that you pay for the brand. And also that I want to buy my car in Germany. They are a lot cheaper than Brazil.